### PR TITLE
[ BB2-1092 ] Create auth flow methods

### PR DESCRIPTION
**Pull request by dtisza1**
**JIRA Ticket:**
[BB2-1092](https://jira.cms.gov/browse/BB2-1092)

**User Story or Bug Summary:**

This adds AUTH related functionality to the BlueButton SDK class and tests.

### What Does This PR Do?

**The following is a summary of the files in this PR**

* Create a separeate file `auth.ts` to include AUTH components of the `BlueButton` class.

* Create PKCE functions in `auth.ts`

* Create exported type and functions to be used as `BlueButton` class methods on `index.ts`:
  - `AuthData` type:
    - Contains a set of auth related data w/ `codeChallenge`, `verifier`, and `state` values.
  - `generateAuthData()`
    - Returns an `AuthData` type.
  - `generateAuthorizeUrl(AuthData)`
    - Generates a new authorization URL string using passed in `AuthData`.

* Add methods from `auth.ts` to `BlueButton` class in `index.ts`.

* Add `callBackUrl` to `BlueButtonCofig` type and `BlueButton` class properties in `index.ts`.

* Update .gitignore for yarn-errors.log

* Add method for generating token post data
  - Move types in auth.ts to types/auth.ts
  - Create method `generateTokenPostData()` taking callback code and state inputs to generate post data for token endpoint call.
  - Create tests for `generateTokenPostData()` return and state does not match exception.

* Add method to validate call back params
  - Add `validateCallbackRequestQueryParams()` function in `src/auth.ts`
  - Add `validateCallbackRequestQueryParams()` function as a method in `src/auth.ts`
  - Create `AuthorizationToken` class in `src/entities/auth.ts`.
  - Create error message ENUMs in `
  - Add tests for `validateCallbackRequestQueryParams()` method in `src/auth.test.ts`
  - Update `index.test.ts` for `callBackUrl`

* Change class/type namings to be Capitalized

* Create get authorization token functionality
  - Add `moment` and `axios` to dependencies in `package.json`
  - Change `AccessToken` to `AuthorizationToken` class and refactor attribute namings in `src/entities/AuthorizationToken.ts`
  - Create `getAuthorizationToken()` async function in `src/auth.ts`.
  - Create `getAuthorizationToken()` public async method for `BlueButton` class `in `src/index.ts` that utilizes underlying async function
  - Add tests related to get authorization token functionality in `src/auth.test.ts`
  - Add `yarn.lock` to `.gitignore` file.

* Create and use `getAccessTokenUrl()` getter method for `BlueButton` class.

* Create and use `getAuthorizationUrl()` getter method for `BlueButton` class.

* Add token endpoint response execptions
  - Add `AUTH_TOKEN_URL_*` error enums to `src/enums/errors.ts`
  - Add private `_authResponseStatusCode` to `BlueButton` class with `getAuthResponseStatusCode()` and `setAuthResponseStatusCode()` methods in `src/index.ts`
  - Add token endpoint response code exceptions for `getAuthorizationToken()` function in `src/auth.ts`
  - Add tests for token endpoint response code exceptions of `getAuthorizationToken()` method in `src/auth.test.ts`

* Update per PR review comments and add validate response tests.
  - Change some namings related to camel case convention
  - Add tests to validate responses in `src/auth.tests.ts`.

* Remove Axios response error handling and tests
  - Any Axios related execptions should be handled by the API user. We should outline the handling in our usage guide documentation.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check these things, in particular:

* Are there any unhandled and/or untested edge cases you can think of?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
* Naming in general? Do the namings of variables, methods, classes, and other items make sense? Are they too verbose, or not verbose enough? Any redundancies? Are the comments descriptive enough?
* Are audit logging items properly handled? Is sensitive information either not included or hashed?

* *What/how to review:*
  - Does the SQL for the views look accurate?
  - I don't think testing is needed for review, but if you want you can go in to AWS Athena to try out the SQL in the query editor. There is some instruction in the readme.md that may be helpful for this.
  - Do the field namings make sense?  Do you have any renaming suggestions?
  - NOTE: As a future task, I will be adding descriptions of key metrics to the Qucksight Dashboard.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls?  **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following pre-requisite changes have been fully deployed:

None.
  
### Any Migrations?

* [ ] Yes, there are migrations
  * [ ] The migrations should be run PRIOR to the code being deployed
  * [ ] The migrations should be run AFTER the code is deployed
  * [ ] There is a more complicated migration plan (downtime, etc) 
* [X] No migrations

### Submitter Checklist

I have gone through and verified that...:

* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,6 +14,7 @@
     "build:types": "tsc -p ./tsconfig.json --outDir build --declaration true && api-extractor run",
     "docs": "typedoc src --out docs",
     "test": "jest",
+    "test:auth": "jest -- auth.test.ts",
```
Comment (nbragdon) [inline package.json:17]: You can actually just add any params you want onto the yarn function and they will get passed through. So for this you can just do: `yarn test auth` and it'll test only this file

Comment (dtisza1) [reply package.json:17]: @nbragdon Cool. Just removed and tried that.

File: src/auth.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+/*
+auth.ts - Provides auth related methods for the Bluebutton class
+*/
+
+import BlueButton from "./index";
```
Comment (nbragdon) [inline src/auth.ts:5]: You can actually just write this as: `import BlueButton from ".";`. Basically `index` is saying that is in the "default" thing to import from that directory

File: src/auth.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import {
+  generateCodeChallenge,
+  generateRandomState,
+} from "./utils/generatePKCE";
+
+export function setCallBackUrl(this: BlueButton, callBackUrl: string): any {
```
Comment (nbragdon) [inline src/auth.ts:11]:
```suggestion
export function setCallBackUrl(this: BlueButton, callBackUrl: string) {
```
For functions that don't return anything, we'll actually want to use `void`. Typescript does a pretty good job of inferring the typing though, so if you don't have a `return` statement, it knows that it is just void. If you are using VSCode, you can hover over function/variable names and it'll show you the inferred typings

File: src/auth.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+export function generateAuthorizeUrl(this: BlueButton): any {
+  const BB2_AUTH_URL = this.baseUrl + "/" + this.version + "/o/authorize";
+
+  let pkceParams = "";
+  this.state = generateRandomState();
+
+  if (this.usePKCE) {
+    this.codeChallenge = generateCodeChallenge();
+    pkceParams =
```
Comment (nbragdon) [inline src/auth.ts:27]: Not sure why I did it like this in the sample app but... we should really use string templates for this, so much cleaner!
```suggestion
    pkceParams = `&code_challenge_method=S256&code_challenge=${this.codeChallenge.codeChallenge}`
```
The syntax is just using the backticks at the start and end. Inside the `${}` you can use variables, operations, etc. as long as the final result in a string or able to be parsed into a string. You can also add new lines for readability wherever you want.

File: src/index.ts [hunk 5]
```diff
@@ -19,6 +24,13 @@ export default class BlueButton {
   version: string;
   baseUrl: string;
 
+  // AUTH props
+  usePKCE: boolean;
```
Comment (nbragdon) [inline src/index.ts:28]: Is there ever a reason to NOT use PKCE? Should we even let that be an option?

File: src/index.ts [hunk 6]
```diff
@@ -19,6 +24,13 @@ export default class BlueButton {
...
+  callBackUrl: string;
+  state: string;
```
Comment (nbragdon) [inline src/index.ts:30]: The state, codeChallenge definitely CANNOT be properties on this top level class. A single instance of this class could be used to authorize any number of beneficiaries, whereas the state/codeChallenge are unique per bene, per auth attempt. So those will be something the 3rd party app code will have to manage/provide to our functions.

File: src/index.ts [hunk 7]
```diff
@@ -19,6 +24,13 @@ export default class BlueButton {
...
+  codeChallenge: CodeChallenge;
+  authorizeUrl: string;
```
Comment (nbragdon) [inline src/index.ts:32]: I don't think this needs to be a property - we can just always generate it.

File: src/utils/generatePKCE.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    .replace(/\+/g, "-")
+    .replace(/\//g, "_")
+    .replace(/=/g, "");
+}
+
+function sha256(str: string): Buffer {
+  return crypto.createHash("sha256").update(str).digest();
+}
+
+export function generateCodeChallenge(): CodeChallenge {
```
Comment (nbragdon) [inline src/utils/generatePKCE.ts:23]: I know this is basically just lifted from the sample app but I feel like it's named a little weird. Just because it turns out to be: `codeChallenge.codeChallenge` sort of thing... So maybe consider a different name for the wrapper around verifier + codeChallenge? Maybe something like pkceData? I dunno...

File: src/auth.ts (hunk 3 above)
Comment (dtisza1) [reply src/auth.ts:11]: @nbragdon Yikes. Knew better... Forgot to remove that. Needed it for the other function, then did a copy to start a new one an didn't remove :man_facepalming: afterward.

File: src/index.ts (hunk 5 above)
Comment (dtisza1) [reply src/index.ts:28]: @nbragdon That's right. We don't want to even give the option of not using it. I'll remove it.

File: src/auth.ts (hunk 2 above)
Comment (nbragdon) [reply src/auth.ts:5]: I've also recently taken to letting my IDE do all the imports for me (and it uses the format I pasted above ^)

File: src/index.ts (hunk 6 above)
Comment (dtisza1) [reply src/index.ts:30]: @nbragdon Per slack question, I see this :-) Wasn't thinking about the next auth parts yet and usage of those. Will update.

File: src/auth.ts (hunk 3 above)
Comment (dtisza1) [reply src/auth.ts]: @nbragdon Related to the `this` usage and having the function code in another file, I like your example better. That is less confusing.  Updating now...

File: src/auth.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
+/*
+auth.ts - Provides auth related methods for the Bluebutton class
+*/
+import crypto from "crypto";
+
+import BlueButton from ".";
+
+export type authData = {
```
Comment (nbragdon) [inline src/auth.ts:8]: I generally see classes/types capitalized 
```suggestion
export type AuthData = {
```

File: src/auth.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    verifier: pkceData.verifier,
+    state: generateRandomState(),
+  };
+}
+
+export function generateAuthorizeUrl(
+  bb: BlueButton,
+  authData: authData
+): string {
+  const BB2_AUTH_URL = bb.baseUrl + "/" + bb.version + "/o/authorize";
```
Comment (nbragdon) [inline src/auth.ts:56]:
```suggestion
  const BB2_AUTH_URL = `${bb.baseUrl}/${bb.version}/o/authorize`;
```

File: src/auth.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    .replace(/\+/g, "-")
+    .replace(/\//g, "_")
+    .replace(/=/g, "");
+}
+
+function sha256(str: string): Buffer {
+  return crypto.createHash("sha256").update(str).digest();
+}
+
+type pkceData = {
```
Comment (nbragdon) [inline src/auth.ts:21]:
```suggestion
type PkceData = {
```

Generally types/classes are capitalized

File: src/types/auth.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
+/* Types for Auth related methods */
+
+export type authData = {
```
Comment (nbragdon) [inline src/types/auth.ts:3]:
```suggestion
export type AuthData = {
```
You can put these in a separate file but personally I tend to just keep them in the same file that they are most correlated to (ie function/class where they are used)

File: src/auth.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+): string {
+  const BB2_AUTH_URL = bb.baseUrl + "/" + bb.version + "/o/authorize";
+
+  const pkceParams = `code_challenge_method=S256&code_challenge=${authData.codeChallenge}`;
+
+  return `${BB2_AUTH_URL}?client_id=${bb.clientId}&redirect_uri=${bb.callBackUrl}&state=${authData.state}&response_type=code&${pkceParams}`;
+}
+
+//  Generates post data for call to access token URL
+export function generateTokenPostData(
```
Comment (nbragdon) [inline src/auth.ts:59]: Hmmm I'm questioning the value add of this function... we might just be able to remove it? I'm also a bit worried about returning sensitive data in the response. Might make it easier for users to accidentally log out something sensitive unintentionally.

File: src/types/auth.ts (hunk 12 above)
Comment (dtisza1) [reply src/types/auth.ts]: @nbragdon Moved these in to auth.ts .

File: src/auth.ts (hunk 13 above)
Comment (dtisza1) [reply src/auth.ts:62]: @nbragdon Was initially thinking we wouldn't be making actual requests via axios in the SDK, so was providing this. Just about to work on the `getAccessToken()` method, so will make this function private to it.

File: src/entities/AccessToken.ts [hunk 14]
```diff
@@ -0,0 +1,33 @@
+import moment from "moment";
+
+export interface IAccessToken {
```
Comment (nbragdon) [inline src/entities/AccessToken.ts:3]:
```suggestion
export type AccessTokenData = {
```
Just an alternate suggestion for this. Might be better to just make this a type instead of an interface.

File: src/entities/AccessToken.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+  access_token: string;
+  expires_in: number;
+  token_type: string;
+  scope: [string];
+  refresh_token: string;
+  patient: string;
+  expires_at?: number;
+}
+
+export default class AccessToken implements IAccessToken {
```
Comment (nbragdon) [inline src/entities/AccessToken.ts:13]:
```suggestion
export default class AccessToken {
```

File: src/entities/AccessToken.ts [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+  public access_token: string;
```
Comment (nbragdon) [inline src/entities/AccessToken.ts:14]:
```suggestion
  public accessToken: string;
```

We could then also name these variable in camel case instead. So it'll take "raw json" from the server as part of the constructor and then turn it into a useful class that provides value.

File: src/entities/AccessToken.ts [hunk 17]
```diff
@@ -0,0 +1,33 @@
+import moment from "moment";
```
Comment (nbragdon) [inline src/entities/AccessToken.ts:1]: If we are going to be using moment in the "shippable" code - we'll need to add it to regular dependencies `yarn add moment` not just the dev dependencies

Comment (dtisza1) [reply src/entities/AccessToken.ts]: @nbragdon Fixed before your comment, but hadn't pushed up yet. Also added axios in the same way.

File: src/entities/AccessToken.ts (hunk 16 above)
Comment (dtisza1) [reply src/entities/AccessToken.ts]: @nbragdon I changed this class back to `AuthorizationToken` in the PR and copied the camel case namings from James's PR. TY!

File: src/entities/AccessToken.ts (hunk 14 above)
Comment (dtisza1) [reply src/entities/AccessToken.ts]: Just updating this. Will push up shortly.

File: src/auth.ts (hunk 10 above)
Comment (dtisza1) [reply src/auth.ts]: Updated this. TY!

File: src/auth.ts (hunk 11 above)
Comment (dtisza1) [reply src/auth.ts]: @nbragdon Updated this and other related type/class vars. Will do another pass to see if there are more naming changes. TY!

File: src/auth.ts (hunk 13 above)
Comment (dtisza1) [reply src/auth.ts:77]: @nbragdon I removed `generateTokenPostData()` and `validateCallbackRequestQueryParams()` from the `BlueButton` class methods.

They are utilized inside of the `getAuthorizationToken()` method now.

File: src/auth.test.ts [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+        "patient/ExplanationOfBenefit.read",
+        "patient/Patient.read",
+        "profile",
+      ],
+      refresh_token: "jV3knA4xmZ1enK8Rg1Lub5hmIsc9Ad",
+      patient: "-20140000000051",
+    },
+  };
+
+  it("expect successful response", async () => {
```
Comment (nbragdon) [inline src/auth.test.ts:115]:
```suggestion
  it("expect successful response", async () => {
    mockedAxios.post.mockResolvedValueOnce(mockResponse);
    const expectedAuthToken = new AuthorizationToken(mockResponse.data);
    const ret = await bb.getAuthorizationToken(
        authData,
        "test-code",
        authData.state,
        "test-error"
      );
      expect(axios.post).toHaveBeenCalledWith(
      BB2_ACCESS_TOKEN_URL,
      expect.anything()
    );
    // write some test to validate `ret`

Comment (nbragdon) [reply src/auth.test.ts:115]: Since you have the `async` at the top, I don't think you need a separate block here. If an error is thrown during a test, the test will fail. It'll also just tell you what error happened.

File: src/auth.test.ts [hunk 19]
```diff
@@ -0,0 +1,219 @@
+import axios from "axios";
+
+jest.mock("axios");
+
+const mockedAxios = axios as jest.Mocked<typeof axios>;
```
Comment (nbragdon) [inline src/auth.test.ts:5]: I don't think you need this here. Was just checking the mock docs - https://jestjs.io/docs/mock-functions. They actually show axios mocking as an example:
```
import axios from 'axios';
import Users from './users';

jest.mock('axios');

test('should fetch users', () => {
  const users = [{name: 'Bob'}];
  const resp = {data: users};
  axios.get.mockResolvedValue(resp);

  // or you could use the following depending on your use case:
  // axios.get.mockImplementation(() => Promise.resolve(resp))

  return Users.all().then(data => expect(data).toEqual(users));
});
```
You can just use `axios` and it'll actually be the mocked object.

File: src/auth.test.ts [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+
+  it("expect successful response with missing expires_at", async () => {
+    mockedAxios.post.mockResolvedValueOnce(mockResponseMissingExpiresAt);
+
+    await expect(async () => {
+      const expectedAuthToken = new AuthorizationToken(
+        mockResponseMissingExpiresAt.data
+      );
+
+      const ret = await bb.getAuthorizationToken(
```
Comment (nbragdon) [inline src/auth.test.ts:145]: I think since `getAuthorizationToken` has a return value, we'll make sure the return value is valid. Checking that axios got the right input is a good test as well but then what the code does with the results of that we'll also need to validate via the return object.

File: src/auth.ts [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+
+  return `${bb.getAuthorizationUrl()}?client_id=${bb.clientId}&redirect_uri=${
+    bb.callbackUrl
+  }&state=${AuthData.state}&response_type=code&${pkceParams}`;
+}
+
+//  Generates post data for call to access token URL
+export function generateTokenPostData(
+  bb: BlueButton,
+  AuthData: AuthData,
```
Comment (nbragdon) [inline src/auth.ts:79]: This is a smaller nitpick but I noticed that `AuthData` wasn't normal camel case: `authData`. Basically class/type/interface definitions should start with a capital. Variables should all start lowercase.

File: src/auth.ts [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+    grant_type: "authorization_code",
+    redirect_uri: bb.callbackUrl,
+    code_verifier: AuthData.verifier,
+    code_challenge: AuthData.codeChallenge,
+  };
+}
+
+function validateCallbackRequestQueryParams(
+  AuthData: AuthData,
+  callbackCode: string | undefined,
```
Comment (nbragdon) [inline src/auth.ts:95]:
```suggestion
  callbackCode?: string,
```
You can use the question mark as a quick way to denote optional

File: src/auth.ts [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+
+    const postData = generateTokenPostData(bb, AuthData, callbackRequestCode);
+
+    const resp = await axios.post(bb.getAccessTokenUrl(), postData);
+
+    const authToken = new AuthorizationToken(resp.data);
+
+    return authToken;
+  } catch (e) {
+    throw e;
```
Comment (nbragdon) [inline src/auth.ts:141]: If you are just going to catch and then throw without doing anything else... what is the purpose of the try catch block?

File: src/index.ts [hunk 24]
```diff
@@ -63,6 +70,7 @@ export default class BlueButton {
     this.callbackUrl = bbJsonConfig.callbackUrl;
     this.clientSecret = bbJsonConfig.clientSecret;
     this.version = bbJsonConfig.version;
+    this.callbackUrl = bbJsonConfig.callbackUrl;
```
Comment (nbragdon) [inline src/index.ts:73]: This already exists above ^

File: src/index.ts [hunk 25]
```diff
@@ -77,4 +85,35 @@ export default class BlueButton {
           : SANDBOX_BASE_URL,
     };
   }
+
+  public getAccessTokenUrl(): string {
+    return this.baseUrl + "/" + this.version + "/o/token/";
```
Comment (nbragdon) [inline src/index.ts:90]:
```suggestion
    return `${this.baseUrl}/${this.version}/o/token/`;
```
Just for consistency sake

File: src/index.ts [hunk 26]
```diff
@@ -77,4 +85,35 @@ export default class BlueButton {
...
+  }
+
+  public getAuthorizationUrl(): string {
+    return `${this.baseUrl}/${this.version}/o/authorize`;
```
Comment (nbragdon) [inline src/index.ts:94]:
```suggestion
    return `${this.baseUrl}/v${this.version}/o/authorize`;
```
Version is just the number, without the v in front.

File: src/auth.test.ts (hunk 18 above)
Comment (dtisza1) [reply src/auth.test.ts:124]: @nbragdon This works. I think did this when trying to get mocks to work. TY!

File: src/auth.test.ts (hunk 19 above)
Comment (dtisza1) [reply src/auth.test.ts:5]: @nbragdon I get this error when using axios directly with out the `mockedAxios`. Found that as a solution researching the error.

```
    src/auth.test.ts:125:16 - error TS2339: Property 'mockResolvedValueOnce' does not exist on type '<T = any, R = AxiosResponse<T>>(url: string, data?: any, config?: AxiosRequestConfig | undefined) => Promise<R>'.

    125     axios.post.mockResolvedValueOnce(mockResponse);
```

Comment (nbragdon) [reply src/auth.test.ts:5]: Making me wonder if we really need typescript on our tests...

File: src/auth.ts [hunk 27]
```diff
@@ -0,0 +1,175 @@
...
+  try {
+    validateCallbackRequestQueryParams(
+      AuthData,
+      callbackRequestCode,
+      callbackRequestState,
+      callbackRequestError
+    );
+
+    const postData = generateTokenPostData(bb, AuthData, callbackRequestCode);
+    const resp = await axios.post(bb.getAccessTokenUrl(), postData);
```
Comment (nbragdon) [inline src/auth.ts:134]: By design, I think axios will thrown an error for any non-200 response. So we can probably just allow that error to surface to the user of the SDK. I think it's probably ok if they know what the underlying library is that is making the requests.

Comment (nbragdon) [reply src/auth.ts:134]: https://axios-http.com/docs/example check out the async/await section and here as well: https://axios-http.com/docs/handling_errors

File: src/entities/AuthorizationToken.ts [hunk 28]
```diff
@@ -0,0 +1,33 @@
+import moment from "moment";
```
Comment (nbragdon) [inline src/entities/AuthorizationToken.ts:1]: This is looking really good to me 👏

File: src/index.ts [hunk 29]
```diff
@@ -21,6 +28,7 @@ export default class BlueButton {
   callbackUrl: string;
   version: string;
   baseUrl: string;
+  private _authResponseStatusCode: number;
```
Comment (nbragdon) [inline src/index.ts:31]: We shouldn't store variables like this on the BB class. This should just hold the high level config for the app to make requests to BB2. It shouldn't be storing anything on any individual auth flow/fhir request. A single class could be authorizing a lot of beneficiaries, fetching a lot of data.

File: src/auth.test.ts (hunk 20 above)
Comment (dtisza1) [reply src/auth.test.ts]: @nbragdon Added some checks for the return on `getAuthorizationToken()`.

File: src/auth.ts (hunk 21 above)
Comment (dtisza1) [reply src/auth.ts]: Updated this.

File: src/auth.ts (hunk 23 above)
Comment (dtisza1) [reply src/auth.ts]: @nbragdon Just removed the try/catch. Had added this at some point due to tests not working quite right. Still works now with this removed. Can't recall what trouble I was having :-)   TY!

File: src/index.ts (hunk 24 above)
Comment (dtisza1) [reply src/index.ts]: I think I missed this when resolving merge conflicts. Removed this duplicate. TY!

File: src/index.ts (hunk 25 above)
Comment (dtisza1) [reply src/index.ts]: Updated this and verified other areas are changed over to this method. TY!

File: src/index.ts (hunk 26 above)
Comment (dtisza1) [reply src/index.ts]: Oops. Added the "v" in for this and other areas. TY!

File: src/auth.ts (hunk 27 above)
Comment (dtisza1) [reply src/auth.ts]: @nbragdon I removed all the Axios response code exceptions and tests.  Axios errors should bubble up to the API user to handle. We can provide SDK documentation on how to do this.

File: src/index.ts (hunk 29 above)
Comment (dtisza1) [reply src/index.ts]: @nbragdon I removed this.

**Review by nbragdon — APPROVED** [review]
Nice job on this 👍

File: src/entities/AuthorizationToken.ts [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+  public scope: string[];
+  public refreshToken: string;
+  public patient: string;
+
+  constructor(authToken: AuthorizationTokenData) {
+    this.accessToken = authToken.access_token;
+    this.expiresIn = authToken.expires_in;
+    this.expiresAt = authToken.expires_at
+      ? authToken.expires_at
+      : moment().add(this.expiresIn).valueOf();
```
Comment (JFU-GIT) [inline src/entities/AuthorizationToken.ts:27]: moment().valueOf() return epoch time in milli-seconds, and this.expiresIn is in seconds (36000), probably need conversion.

**Review by JFU-GIT — CHANGES_REQUESTED** [review]
moment().valueOf() return epoch time in milli-seconds, and this.expiresIn is in seconds (36000), probably need conversion.

File: src/entities/AuthorizationToken.ts (hunk 30 above)
Comment (dtisza1) [reply src/entities/AuthorizationToken.ts:27]: @JFU-GIT Good catch!  TY!   Will update this to be:

```
      : moment().add(this.expiresIn * 1000).valueOf();

```

File: src/auth.ts [hunk 31]
```diff
@@ -0,0 +1,143 @@
...
+  };
+}
+
+export function generateAuthorizeUrl(
+  bb: BlueButton,
+  AuthData: AuthData
+): string {
+  const pkceParams = `code_challenge_method=S256&code_challenge=${AuthData.codeChallenge}`;
+
+  return `${bb.getAuthorizationUrl()}?client_id=${bb.clientId}&redirect_uri=${
```
[the remark this answers is not in the document]
Comment (dtisza1) [reply src/auth.ts:71]: We should only allow the use of PKCE and state values in our SDK's. We should not allow the SDK users to disable this.
